Bearded.UI.Tests/Controls/FocusTests.cs

```diff
@@ -0,0 +1,429 @@
+using System;
+using Bearded.UI.Controls;
+using Bearded.UI.Rendering;
+using FluentAssertions;
+using Xunit;
+
+namespace Bearded.UI.Tests.Controls
+{
+    public static class FocusTests
+    {
+        public class InitialState
+        {
+            [Fact]
+            public void IsFocusedIsFalse()
+            {
+                var (_, child) = rootWithFocusableChild();
+
+                child.IsFocused.Should().BeFalse();
+            }
+
+            [Fact]
+            public void RootHasFocusedDescendantIsFalse()
+            {
+                var (root, _) = rootWithFocusableChild();
+
+                root.HasFocusedDescendant.Should().BeFalse();
+            }
+
+            [Fact]
+            public void FocusManagedFocusedControlIsNull()
+            {
+                var (root, _) = rootWithFocusableChild();
+
+                root.FocusManager.FocusedControl.Should().BeNull();
+            }
+        }
+
+        public static class Focus
+        {
+            public class OnUnfocusedControlWithoutExistingFocusedControl
+            {
+                [Fact]
+                public void SetsIsFocusedTrue()
+                {
+                    var (_, child) = rootWithFocusableChild();
+
+                    child.Focus();
+
+                    child.IsFocused.Should().BeTrue();
+                }
+
+                [Fact]
+                public void SetsFocusManagerFocusedControl()
+                {
+                    var (root, child) = rootWithFocusableChild();
+
+                    child.Focus();
+
+                    root.FocusManager.FocusedControl.Should().Be(child);
+                }
+
+                [Fact]
+                public void CallsFocusedMethod()
+                {
+                    var (_, child) = rootWithFocusableChild();
+
+                    child.Focus();
+
+                    child.FocusedMethodCalled.Should().BeTrue();
+                }
+
+                [Fact]
+                public void ThrowsIfControlCanNotBeFocused()
+                {
+                    var (_, child) = rootWithNonFocusableChild();
+
+                    Action action = () => child.Focus();
+
+                    action.Should().Throw<InvalidOperationException>();
+                }
+
+                [Fact]
+                public void SetsHasFocusedDescendantTrueOnParent()
+                {
+                    var (_, intermediate, child) = rootWithNestedFocusableChild();
+
+                    child.Focus();
+
+                    intermediate.HasFocusedDescendant.Should().BeTrue();
+                }
+
+                [Fact]
+                public void SetsHasFocusedDescendantTrueOnRoot()
+                {
+                    var (root, child) = rootWithFocusableChild();
+
+                    child.Focus();
+
+                    root.HasFocusedDescendant.Should().BeTrue();
```

Comment:
This test would be more meaningful if there was an intermediate I think to ensure it works for more than just the immediate parent.